modules/33acrossAnalyticsAdapter.js

```diff
@@ -0,0 +1,475 @@
+    auctions: [ auctions[completedAuctionId] ],
+  }
+}
+
+function getBidsForTransaction(auctionId, transactionId) {
+  const auction = locals.cache.auctions[auctionId];
+  return auction.adUnits.find(adUnit => adUnit.transactionId === transactionId).bids;
+}
+
+function getBid(auctionId, bidId) {
+  const auction = locals.cache.auctions[auctionId];
+  for (let adUnit of auction.adUnits) {
+    for (let bid of adUnit.bids) {
+      if (bid.bidId === bidId) {
+        return bid;
+      }
+    }
+  }
+}
+
+/**
+ * @param {Object} args
+ * @param {Object} args.args Event data
+ * @param {EVENTS[keyof EVENTS]} args.eventType
+ */
+function analyticEventHandler({ eventType, args }) {
+  if (!locals.cache) {
+    log.error('Something went wrong. Analytics cache is not initialized.');
+    return;
+  }
+
+  switch (eventType) {
+    case EVENTS.AUCTION_INIT:
+      onAuctionInit(args);
+      break;
+    case EVENTS.BID_REQUESTED:
+      onBidRequested(args);
+      break;
+    case EVENTS.BID_RESPONSE:
+      onBidResponse(args);
+      break;
+    case EVENTS.BID_WON:
+      onBidWon(args);
+      break;
+    case EVENTS.AUCTION_END:
+      onAuctionEnd(args);
+      break;
+    default:
+      break;
+  }
+}
+
+function onAuctionInit({ adUnits, auctionId, bidderRequests }) {
+  if (typeof auctionId !== 'string' || !Array.isArray(bidderRequests)) {
+    log.error('Analytics adapter failed to parse auction.');
+  }
+
+  locals.cache.auctions[auctionId] = {
+    auctionId,
+    adUnits: adUnits.map(au => {
+      return {
+        transactionId: au.transactionId,
+        adUnitCode: au.code,
+        // Note: GPID supports adUnits that have matching `code` values by appending a `#UNIQUIFIER`.
+        // The value of the UNIQUIFIER is likely to be the div-id,
+        // but, if div-id is randomized / unavailable, may be something else like the media size)
+        slotId: deepAccess(au, 'ortb2Imp.ext.gpid') || deepAccess(au, 'ortb2Imp.ext.data.pbadslot', au.code),
+        mediaTypes: Object.keys(au.mediaTypes),
+        sizes: au.sizes.map(size => size.join('x')),
+        bids: [],
+      }
+    }),
+    userIds: Object.keys(deepAccess(bidderRequests, '0.bids.0.userId', {})),
+  };
+
+  locals.transactionManagers[auctionId] ||=
+    new TransactionManager({
+      timeout: analyticsAdapter.getTimeout(),
+      onComplete() {
+        sendReport(
+          createReportFromCache(locals.cache, auctionId),
+          analyticsAdapter.getUrl()
+        );
+        delete locals.transactionManagers[auctionId];
+      }
+    });
+}
+
+function onBidRequested({ auctionId, bids }) {
+  for (let { bidder, bidId, transactionId, src } of bids) {
+    getBidsForTransaction(auctionId, transactionId).push({
+      bidder,
+      bidId,
+      status: 'pending',
+      hasWon: 0,
+      source: src,
+    });
+
+    // if there is no manager for this auction, then the auction has already been completed
+    // eslint-disable-next-line no-unused-expressions
```

Comment:
Why does `eslint` think there's an unused expression? Does it have trouble with the optional chaining operator?

Reply:
Yes, the optional chaining operator is an issue for the particular version of gulp-eslint that ships with prebid. I'm going to update this to use a guard operator instead.